### PR TITLE
implement isrmatrix_vector_product

### PR DESCRIPTION
**Pull request by apapapostolou**
Implement and test the matrix-vector product of an ISR one-particle operator through second order in perturbation theory. Both symmetric and asymmetric one-particle operators are allowed.

### CODE REVIEW

**Comment by maxscheurer** [comment]
Thanks for the contribution. I think the implementation should be in line with the `AdcMatrix` code structure. So one would need to re-group the equations into blocks and then basically copy/paste into the scaffold of the `AdcMatrix` code. I guess it won't be too much work ☺️

File: adcc/IsrMatrix.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env python3
+## vi: tabstop=4 shiftwidth=4 softtabstop=4 expandtab
+## ---------------------------------------------------------------------
+##
+## Copyright (C) 2018 by the adcc authors
```
Comment (maxscheurer) [inline adcc/IsrMatrix.py:5]:
```suggestion
## Copyright (C) 2022 by the adcc authors
```

File: adcc/IsrMatrix.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        Initialise an ISR matrix of a given one-particle operator
+        for the provided ADC method.
+
+        Parameters
+        ----------
+        method : str or adcc.AdcMethod
+            Method to use.
+        hf_or_mp : adcc.ReferenceState or adcc.LazyMp
+            HF reference or MP ground state.
+        dips : adcc.OneParticleOperator or list of adcc.OneParticleOperator objects
```
Comment (maxscheurer) [inline adcc/IsrMatrix.py:55]: I suggest to rename this to `operators`.

File: adcc/IsrMatrix.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                            "either a LazyMp, a ReferenceState or a "
+                            "HartreeFockSolution_i.")
+
+        if not isinstance(method, AdcMethod):
+            method = AdcMethod(method)
+
+        if not isinstance(dips, list):
+            self.dips = [dips]
+        else:
+            self.dips = dips
```
Comment (maxscheurer) [inline adcc/IsrMatrix.py:75]: Please check that dips/operators are instances of OneParticleOperator.

File: adcc/adc_pp/bmatrix.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+`apply` is a function mapping an AmplitudeVector to the contribution of this
+block to the result of applying the ADC matrix.
+"""
+AdcBlock = namedtuple("AdcBlock", ["apply"])
+
+
+def block(ground_state, dip, spaces, order, variant=None):
+    """
+    Gets ground state, one-particle matrix elements associated
+    with the dipole operator, spaces (ph, pphh and so on)
```
Comment (maxscheurer) [inline adcc/adc_pp/bmatrix.py:46]:
```suggestion
    with a one-particle operator, spaces (ph, pphh and so on)
```

File: adcc/adc_pp/bmatrix.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env python3
+## vi: tabstop=4 shiftwidth=4 softtabstop=4 expandtab
+## ---------------------------------------------------------------------
+##
+## Copyright (C) 2020 by the adcc authors
```
Comment (maxscheurer) [inline adcc/adc_pp/bmatrix.py:5]:
```suggestion
## Copyright (C) 2022 by the adcc authors
```

File: adcc/IsrMatrix.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            if all(dv.dot(dv) < 1e-12 for dv in diffv):
+                return [
+                    AmplitudeVector(ph=-1.0 * mv.ph, pphh=-1.0 * mv.pphh)
+                    for mv in self.matvec(v)
+                ]
+            # operators without any symmetry
+            else:
+                return NotImplemented
+
+    def __matmul__(self, other):
```
Comment (maxscheurer) [inline adcc/IsrMatrix.py:152]: Can you please add a docstring about what happens when you apply the matrix with multiple operators to multiple vectors? Do I get a 2D list with N_operators x N_vectors back?

File: adcc/test_IsrMatrix.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python3
+## vi: tabstop=4 shiftwidth=4 softtabstop=4 expandtab
+## ---------------------------------------------------------------------
+##
+## Copyright (C) 2020 by the adcc authors
```
Comment (maxscheurer) [inline adcc/test_IsrMatrix.py:5]:
```suggestion
## Copyright (C) 2022 by the adcc authors
```

**Review by maxscheurer — COMMENTED** [review]
This looks really good! Just a few comments, tests look good to me. After addressing the comments, we can merge this 🚀

File: adcc/IsrMatrix.py (hunk 6 above)
Comment (apapapostolou) [reply adcc/IsrMatrix.py:162]: Done! No, you get a 2D list with N_vectors x N_operators back.

**Review by maxscheurer — APPROVED** [review]
This looks great! 👍 
You want to give this another look, @mfherbst?

**Comment by maxscheurer** [comment]
@apapapostolou It would be great if you could try to update the [TPA](https://github.com/adc-connect/adcc/blob/e58bc07cee2a47fb5967ca46e50a66c27ef9bda9/examples/water/tpa.py) example using your new `IsrMatrix` instead of the transition moments?

**Comment by maxscheurer** [comment]
Thanks a lot for updating the example @apapapostolou 🚀

**Comment by maxscheurer** [comment]
Is this ready from your side, @apapapostolou? 😄

**Comment by apapapostolou** [comment]
> Is this ready from your side, @apapapostolou? 😄

Yes, it is 😄

**Comment by maxscheurer** [comment]
@apapapostolou feel free to create a PR and add yourself to the contributors/authors! 😬

**Pull request merged**